packages/stream_chat_flutter/lib/src/attachment/audio/audio_wave_slider.dart

```diff
@@ -0,0 +1,248 @@
+                        ?.call(details.localPosition.dx / constraints.maxWidth);
+
+                    setState(() {
+                      _dragging = true;
+                    });
+                  },
+                  onHorizontalDragEnd: (details) {
+                    widget.onChangeEnd?.call();
+
+                    setState(() {
+                      _dragging = false;
+                    });
+                  },
+                  onHorizontalDragUpdate: (details) {
+                    widget.onChanged?.call(
+                      min(
+                        max(details.localPosition.dx / constraints.maxWidth, 0),
+                        1,
+                      ),
+                    );
+                  },
+                ),
+              ],
+            );
+          },
+        );
+      },
+    );
+  }
+}
+
+class _AudioBarsPainter extends CustomPainter {
+  _AudioBarsPainter({
+    required this.bars,
+    required this.progressPercentage,
+    this.spacingRatio = 0.01,
+    this.colorLeft = Colors.blueAccent,
+    this.colorRight = Colors.grey,
+    this.barHeightRatio = 1,
+    this.padding = 20,
+  });
+
+  final List<double> bars;
+  final Color colorRight;
+  final Color colorLeft;
+  final double progressPercentage;
+  final double spacingRatio;
+  final double barHeightRatio;
+  final int padding;
+
+  /// barWidth should include spacing, not only the width of the bar.
+  /// progressX should be the middle of the moving button of the slider, not
+  /// initial X position.
+  Color _barColor(double buttonCenter, double progressX) {
+    return (progressX > buttonCenter) ? colorLeft : colorRight;
+  }
+
+  double _barHeight(double barValue, totalHeight) {
+    return max(barValue * totalHeight * barHeightRatio, 2);
+  }
+
+  double _progressToWidth(double totalWidth, double progress) {
+    final availableWidth = totalWidth;
+
+    return availableWidth * progress + padding;
+  }
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    final totalWidth = size.width - padding * 2;
+
+    final spacingWidth = totalWidth * spacingRatio;
+    final totalBarWidth = totalWidth - spacingWidth * (bars.length - 1);
+    final barWidth = totalBarWidth / bars.length;
+    final barY = size.height / 2;
+
+    bars.forEachIndexed((i, barValue) {
+      final barHeight = _barHeight(barValue, size.height);
+      final barX = i * (barWidth + spacingWidth) + barWidth / 2 + padding;
+
+      final rect = RRect.fromRectAndRadius(
+        Rect.fromCenter(
+          center: Offset(barX, barY),
+          width: barWidth,
+          height: barHeight,
+        ),
+        const Radius.circular(50),
+      );
+
+      final paint = Paint()
+        ..color = _barColor(
+          barX + barWidth / 2,
+          _progressToWidth(totalWidth, progressPercentage),
+        );
+      canvas.drawRRect(rect, paint);
+    });
+  }
+
+  @override
+  bool shouldRepaint(covariant CustomPainter oldDelegate) {
+    return true;
```

Comment:
I just tried to optimise the repaint logic, but it lost a lot or performance D= (many frame are lost). 

I believe this happens because this Widget has to redraw itself many times per seconds, than any logic that we put here will be evaluated too ofter, which gets too heavy for the UI.

Reply:
Take a look at this version: 
```
@override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    if (oldDelegate is _AudioBarsPainter) {
      return oldDelegate.bars != bars ||
          oldDelegate.colorRight != colorRight ||
          oldDelegate.colorLeft != colorLeft ||
          oldDelegate.barHeightRatio != barHeightRatio ||
          oldDelegate.spacingRatio != spacingRatio ||
          oldDelegate.inverse != inverse;
    }

    return true;
  }
```

And this:

```
bool shouldRepaint(covariant CustomPainter oldDelegate) {
    if (oldDelegate is _AudioBarsPainter) {
      return oldDelegate.bars != bars;
    }

    return true;
  }
```

Both didn't present a good result